protocols/kad/src/query/peers/closest/disjoint.rs

```diff
@@ -0,0 +1,545 @@
+        // to query the previously queried iterator last.
+        let iter_order = {
+            let mut all = (0..self.iters.len()).collect::<Vec<_>>();
+            let mut next_up = all.split_off(self.last_quiered + 1);
+            next_up.append(&mut all);
+            next_up
+        };
+
+        for i in &mut iter_order.into_iter() {
+            self.last_quiered = i;
+            let iter = &mut self.iters[i];
+
+            loop {
+                match iter.next(now) {
+                    PeersIterState::Waiting(None) => {
+                        match state {
+                            PeersIterState::Waiting(Some(_)) => {
+                                // [`ClosestDisjointPeersIter::next`] returns immediately once a
+                                // [`ClosestPeersIter`] yielded a peer. Thus this state is
+                                // unreachable.
+                                unreachable!();
+                            },
+                            PeersIterState::Waiting(None) => {}
+                            PeersIterState::WaitingAtCapacity => {
+                                state = PeersIterState::Waiting(None)
+                            }
+                            PeersIterState::Finished => state = PeersIterState::Waiting(None),
+                        };
+
+                        break;
+                    }
+                    PeersIterState::Waiting(Some(peer)) => {
+                        if self.yielded_peers.contains_key(&*peer) {
+                            // Another iterator already returned this peer. S/Kademlia requires each
+                            // peer to be only used on one path. Marking it as failed for this
+                            // iterator, asking it to return another peer in the next loop
+                            // iteration.
+                            let peer = peer.into_owned();
+                            iter.on_failure(&peer);
+                        } else {
+                            self.yielded_peers.insert(peer.clone().into_owned(), i);
+                            return PeersIterState::Waiting(Some(Cow::Owned(peer.into_owned())));
+                        }
+                    }
+                    PeersIterState::WaitingAtCapacity => {
+                        match state {
+                            PeersIterState::Waiting(Some(_)) => {
+                                // [`ClosestDisjointPeersIter::next`] returns immediately once a
+                                // [`ClosestPeersIter`] yielded a peer. Thus this state is
+                                // unreachable.
+                                unreachable!();
+                            },
+                            PeersIterState::Waiting(None) => {}
+                            PeersIterState::WaitingAtCapacity => {}
+                            PeersIterState::Finished => state = PeersIterState::WaitingAtCapacity,
```

Comment:
> but the following properties should hold, right?

Correct.

> Would it be feasible to add property tests to that end?

Were you thinking of a mocked `ClosestPeersIter` or an `Arbitrary` implementation for `ClosestPeersIter` that returns `ClosestPeersIter`s in different states?

Reply:
I didn't have a concrete approach in mind, just saying that given it is not easily checked that these properties hold by looking at the implementation, it may be good to have these tested.